backend/testing/tester.py

```diff
@@ -136,23 +136,45 @@ def validate_message(self, body, message):
 
         # compare expected message with received one. exit on
         # mismatch.
-        for key in set(event.keys()).intersection(expected.keys()):
-            if "time" in key:
-                expected[key] += self.time_now
-            assert event[key] == expected[key] or (
-                isinstance(event[key], (list, set))
-                and set(event[key]) == set(expected[key])
-            ), (
-                'Test "{}" - Batch #{} - Type {}: Unexpected'
-                ' value for key "{}". Received: {}, Expected: {}'.format(
+        if isinstance(expected, dict):
+            for key in set(event.keys()).intersection(expected.keys()):
+                if "time" in key:
+                    expected[key] += self.time_now
+                assert event[key] == expected[key] or (
+                    isinstance(event[key], (list, set))
+                    and set(event[key]) == set(expected[key])
+                ), (
+                    'Test "{}" - Batch #{} - Type {}: Unexpected'
+                    ' value for key "{}". Received: {}, Expected: {}'.format(
+                        self.curr_test,
+                        self.curr_idx,
+                        message.delivery_info["routing_key"],
+                        key,
+                        event[key],
+                        expected[key],
+                    )
+                )
+        elif isinstance(expected, list):
+            correct = [True] * len(expected)
+            for idx, expected_item in enumerate(expected):
+                for key in set(event.keys()).intersection(expected_item.keys()):
+                    if "time" in key:
+                        expected_item[key] += self.time_now
+                    if event[key] != expected_item[key] or (
+                        isinstance(event[key], (list, set))
+                        and set(event[key]) != set(expected_item[key])
+                    ):
+                        correct[idx] = False
+                        break
+            if not any(correct):
+                assert 'Test "{}" - Batch #{} - Type {}: Unexpected'
+                " value for complex scenario. Event: {}, Expected List: {}".format(
                     self.curr_test,
                     self.curr_idx,
                     message.delivery_info["routing_key"],
-                    key,
```

Comment:
If possible try to carry the event and expected keys too (so that we know exactly what fails, otherwise debugging is very hard)

Reply:
e.g., besides correct[idx] carry also the failed key.

Reply:
we can't know what failed because there will be multiple fails. which one is the correct fail?

Reply:
assume in-order delivery in the list. If sth does not match, the test should fail (and we should know exactly what failed)